internal/cloud/gcp/cloudbuild.go

```diff
@@ -0,0 +1,224 @@
+	// Logs file names will be of the format `${logs_bucket}/log-${build_id}.txt`
+	logBucket := imageBuild.LogsBucket
+	logBucket = strings.TrimPrefix(logBucket, "gs://")
+	// logBucket may can contain directory in its name if set to a custom value
+	var logObjectDir string
+	if strings.Contains(logBucket, "/") {
+		ss := strings.SplitN(logBucket, "/", 2)
+		logBucket = ss[0]
+		logObjectDir = fmt.Sprintf("%s/", ss[1])
+	}
+	logObject := fmt.Sprintf("%slog-%s.txt", logObjectDir, buildID)
+
+	// Read the log
+	logBuilder := new(strings.Builder)
+	rd, err := storageClient.Bucket(logBucket).Object(logObject).NewReader(ctx)
+	if err != nil {
+		return "", fmt.Errorf("failed to create a new Reader for object '%s/%s': %v", logBucket, logObject, err)
+	}
+	_, err = io.Copy(logBuilder, rd)
+	if err != nil {
+		return "", fmt.Errorf("reading data from object '%s/%s' failed: %v", logBucket, logObject, err)
+	}
+
+	return logBuilder.String(), nil
+}
+
+// CloudbuildBuildCleanup parses the logs for the specified Build job and tries to clean up all resources
+// which were created as part of the job. It returns list of strings with all resources that were deleted
+// as a result of calling this method.
+//
+// Uses:
+//	- Storage API
+//	- Cloud Build API
+//	- Compute Engine API (indirectly)
+func (g *GCP) CloudbuildBuildCleanup(ctx context.Context, buildID string) ([]string, error) {
+	var deletedResources []string
+
+	storageClient, err := storage.NewClient(ctx, option.WithCredentials(g.creds))
+	if err != nil {
+		return deletedResources, fmt.Errorf("failed to get Storage client: %v", err)
+	}
+	defer storageClient.Close()
+
+	buildLog, err := g.CloudbuildBuildLog(ctx, buildID)
+	if err != nil {
+		return deletedResources, fmt.Errorf("failed to get log for build ID '%s': %v", buildID, err)
+	}
+
+	resources, err := cloudbuildResourcesFromBuildLog(buildLog)
+	if err != nil {
+		return deletedResources, fmt.Errorf("extracting created resources from build log failed: %v", err)
+	}
+
+	// Delete all Compute Engine instances
+	for _, instance := range resources.computeInstances {
+		err = g.ComputeInstanceDelete(ctx, resources.zone, instance)
+		if err == nil {
+			deletedResources = append(deletedResources, fmt.Sprintf("instance: %s (%s)", instance, resources.zone))
+		}
+	}
+
+	// Deleting instances in reality takes some time. Deleting a disk while it is still used by the instance, will fail.
+	// Iterate over the list of instances and wait until they are all deleted.
+	for _, instance := range resources.computeInstances {
+		for {
+			_, err := g.ComputeInstanceGet(ctx, resources.zone, instance)
+			// Getting the instance information failed, it is ideleted.
+			if err != nil {
+				break
+			}
+			time.Sleep(1 * time.Second)
+		}
+	}
+
+	// Delete all Compute Engine Disks
+	// Deleting Compute instances takes time, therefore it may happen that deleting the disk will fail, because
```

Comment:
This shouldn't happen if the above loop succeeded, right?

Reply:
Yes. This is mostly FYI, but I should probably remove it from here, because the same information is already included in the comment above the previous loop.

Reply:
fixed